cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -1259,6 +1259,90 @@ struct start_offset_output_iterator {
   }
 };
 
+struct flat_column_num_rows {
+  gpu::PageInfo const* pages;
+  gpu::ColumnChunkDesc const* chunks;
+
+  __device__ size_type operator()(size_type pindex)
+  {
+    gpu::PageInfo const& page = pages[pindex];
+    // ignore dictionary pages and pages belonging to any column containing repetition (lists)
+    if ((page.flags & gpu::PAGEINFO_FLAGS_DICTIONARY) ||
+        (chunks[page.chunk_idx].max_level[gpu::level_type::REPETITION] > 0)) {
+      return 0;
+    }
+    return page.num_rows;
+  }
+};
+
+struct row_counts_nonzero {
+  __device__ bool operator()(size_type count) { return count > 0; }
+};
+
+struct row_counts_different {
+  size_type const expected;
+  __device__ bool operator()(size_type count) { return (count != 0) && (count != expected); }
+};
+
+/**
+ * @brief Detects malformed parquet input data.
+ *
+ * We have seen cases where parquet files can be oddly malformed. This function specifically
+ * detects one case in particular:
+ *
+ * - When when you have a file containing N rows
+ * - For some reason, the sum total of the number of rows over all pages for a given column
+ *   is != N
+ *
+ * @param pages All pages to be decoded
+ * @param chunks Chunk data
+ * @param page_keys Keys (schema id) associated with each page, sorted by column
+ * @param page_index Page indices for iteration, sorted by column
+ * @param expected_row_count Expected row count, if applicable
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ */
+void detect_malformed_pages(hostdevice_vector<gpu::PageInfo>& pages,
+                            hostdevice_vector<gpu::ColumnChunkDesc> const& chunks,
+                            device_span<const int> page_keys,
+                            device_span<const int> page_index,
+                            std::optional<size_t> expected_row_count,
+                            rmm::cuda_stream_view stream)
+{
+  // sum row counts for all non-dictionary, non-list columns. other columns will indicated as 0
+  rmm::device_uvector<int> row_counts(pages.size(), stream);  // worst case:  num keys == num pages
+  auto size_iter = thrust::make_transform_iterator(
+    page_index.begin(), flat_column_num_rows{pages.device_ptr(), chunks.device_ptr()});
+  auto row_counts_begin                     = row_counts.begin();
+  [[maybe_unused]] auto [_, row_counts_end] = thrust::reduce_by_key(rmm::exec_policy(stream),
+                                                                    page_keys.begin(),
+                                                                    page_keys.end(),
+                                                                    size_iter,
+                                                                    thrust::make_discard_iterator(),
+                                                                    row_counts_begin);
```

Comment:
This can just be `auto const row_count_end = thrust::reduce_by_key(...).second;`.